src/main/java/net/glowstone/block/state/StateSerialization.java

```
@@ -0,0 +1,85 @@
+package net.glowstone.block.state;
+
+import net.glowstone.block.state.impl.WoolStateDataReader;
+import org.bukkit.DyeColor;
+import org.bukkit.Material;
+import org.bukkit.material.MaterialData;
+
+import java.util.HashMap;
+import java.util.Map;
+
+public class StateSerialization {
+
+    private static final Map<Material, BlockStateReader> READERS = new HashMap<>();
+
+    public static BlockStateData parse(Material material, String state) throws InvalidBlockStateException {
+        if (state == null || state.trim().isEmpty() || state.trim().equals("*")) {
+            return new BlockStateData();
+        }
+        BlockStateReader reader = getReader(material);
+        if (material == null || reader == null) {
+            throw new InvalidBlockStateException(material, state);
+        }
+        BlockStateData data = new BlockStateData();
+        String[] split = state.trim().split(",");
```

This is a case where I'd say a small Regex could make the expected syntax a bit clearer, and reduce the number of cases you need to check to throw InvalidBlockStateException. Looks like the regex you want is `"^\\s*((\\w+)\\s*=\\s*(\\w+))\\s*(,\\s*(\\w+)\\s*=\\s*(\\w+)\\s*)*$"`. Looks intimidating, yes, but that is just keeping compatibility with the trimmed code above.

To use this regex, you'd be iterating through all the groups (starting at index 1 because the group at index 0 *always* matches the entire string), but taking only the second and third group out of every three, as the first will represent the entire "key=value" match.

---

I hadn't thought of using a regex here, I will consider it. My only concern would be readability, but that can be resolved with a comment.

---

After checking, the regular expression you provided doesn't work all the time. For the string `a=b,c=d,e=f`:

Full match | `a=b,c=d,e=f`
-- | --
Group 1 | `a=b`
Group 2 | `a`
Group 3 | `b`
Group 4 | `,e=f`
Group 5 | `e`
Group 6 | `f`

This would skip the middle section.